calico_cni_k8s_test.go

```diff
@@ -938,6 +939,369 @@ var _ = Describe("CalicoCni", func() {
+					Expect(err).NotTo(HaveOccurred())
+
+					clientset, err := kubernetes.NewForConfig(config)
+					Expect(err).NotTo(HaveOccurred())
+
+					// Now create a K8s pod passing in more than one IPv4 address.
+					name := fmt.Sprintf("run%d-ip", rand.Uint32())
+					pod, err := clientset.Pods(K8S_TEST_NS).Create(&v1.Pod{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: name,
+						},
+						Spec: v1.PodSpec{Containers: []v1.Container{{
+							Name:  fmt.Sprintf("container-%s", name),
+							Image: "ignore",
+						}}},
+					})
+					Expect(err).NotTo(HaveOccurred())
+
+					logger.Infof("Created POD object: %v", pod)
+
+					_, netnspath, _, _, contAddresses, _, _, _, err := CreateContainer(netconfCalicoIPAM, name, "")
+					Expect(err).NotTo(HaveOccurred())
+
+					podIP := contAddresses[0].IP
+					logger.Infof("All container IPs: %v", contAddresses)
+					logger.Infof("Container got IP address: %s", podIP)
+
+					// Pod should have gotten an IP from the second IP Pool.
+					Expect(ipPoolCIDR2.Contains(podIP)).To(BeTrue())
+
+					// Delete the container.
+					_, err = DeleteContainer(netconfCalicoIPAM, netnspath, name)
+					Expect(err).ShouldNot(HaveOccurred())
+				})
+			})
+
+			Context("specify an IPv4 pool in the CNI config, and request a specific IP with ipAddrs annotation from a different IP Pool", func() {
+				It("pod should get the IP specifically requested by the ipAddrs annotation", func() {
+					ipv4pool := "10.0.0.0/24"
+					ipv4pool2 := "20.0.0.0/24"
+					netconfCalicoIPAM := fmt.Sprintf(`
+						{
+						"cniVersion": "%s",
+						"name": "net13",
+						"type": "calico",
+						"etcd_endpoints": "http://%s:2379",
+						"ipam": {
+							"type": "calico-ipam",
+							"ipv4_pools": ["%s"]
+							},
+						"kubernetes": {
+							  "k8s_api_root": "http://127.0.0.1:8080"
+							 },
+						"policy": {"type": "k8s"},
+						"log_level":"info"
+						}`, cniVersion, os.Getenv("ETCD_IP"), ipv4pool)
+
+					// Create a new ipPool.
+					c, _ := client.NewFromEnv()
+
+					// Create the IP Pool specified in the CNI config.
+					testutils.CreateNewIPPool(*c, ipv4pool, false, false, true)
+
+					// Create another IP Pool from which we are going to request a specific IP address.
+					testutils.CreateNewIPPool(*c, ipv4pool2, false, false, true)
+
+					// Specific IP to assign to the pod.
+					requestedIP := net.IPv4(20, 0, 0, 111).To4()
+
+					config, err := clientcmd.DefaultClientConfig.ClientConfig()
+					Expect(err).NotTo(HaveOccurred())
+
+					clientset, err := kubernetes.NewForConfig(config)
+					Expect(err).NotTo(HaveOccurred())
+
+					// Now create a K8s pod passing in more than one IPv4 address.
+					name := fmt.Sprintf("run%d-ip", rand.Uint32())
+					pod, err := clientset.Pods(K8S_TEST_NS).Create(&v1.Pod{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: name,
+							Annotations: map[string]string{
+								"cni.projectcalico.org/ipAddrs": "[\"20.0.0.111\"]",
```

Comment:
Can you generate that IP string from `requestedIP`, instead of duplicating a hardcoded value?